pkg/strategy/atrpin/strategy.go

```diff
@@ -0,0 +1,215 @@
+	}
+
+	if s.Interval == "" {
+		s.Interval = types.Interval5m
+	}
+
+	return nil
+}
+
+func (s *Strategy) Run(ctx context.Context, orderExecutor bbgo.OrderExecutor, session *bbgo.ExchangeSession) error {
+	s.Strategy = &common.Strategy{}
+	s.Strategy.Initialize(ctx, s.Environment, session, s.Market, ID, s.InstanceID())
+
+	atr := session.Indicators(s.Symbol).ATR(s.Interval, s.Window)
+
+	session.MarketDataStream.OnKLineClosed(types.KLineWith(s.Symbol, s.Interval, func(k types.KLine) {
+		if err := s.Strategy.OrderExecutor.GracefulCancel(ctx); err != nil {
+			log.WithError(err).Error("unable to cancel open orders...")
+		}
+
+		account, err := session.UpdateAccount(ctx)
+		if err != nil {
+			log.WithError(err).Error("unable to update account")
+			return
+		}
+
+		baseBalance, _ := account.Balance(s.Market.BaseCurrency)
+		quoteBalance, _ := account.Balance(s.Market.QuoteCurrency)
+
+		lastAtr := atr.Last(0)
+		log.Infof("atr: %f", lastAtr)
+
+		// protection
+		if lastAtr <= k.High.Sub(k.Low).Float64() {
+			lastAtr = k.High.Sub(k.Low).Float64()
+		}
+
+		priceRange := fixedpoint.NewFromFloat(lastAtr * s.Multiplier)
+
+		// if the atr is too small, apply the price range protection with 10%
+		// priceRange protection 10%
+		priceRange = fixedpoint.Max(priceRange, k.Close.Mul(fixedpoint.NewFromFloat(0.1)))
+		log.Infof("priceRange: %f", priceRange.Float64())
+
+		ticker, err := session.Exchange.QueryTicker(ctx, s.Symbol)
+		if err != nil {
+			log.WithError(err).Error("unable to query ticker")
+			return
+		}
+
+		log.Info(ticker.String())
+
+		bidPrice := fixedpoint.Max(ticker.Buy.Sub(priceRange), s.Market.TickSize)
+		askPrice := ticker.Sell.Add(priceRange)
+
+		bidQuantity := s.QuantityOrAmount.CalculateQuantity(bidPrice)
+		askQuantity := s.QuantityOrAmount.CalculateQuantity(askPrice)
+
+		var orderForms []types.SubmitOrder
+
+		position := s.Strategy.OrderExecutor.Position()
+		if !position.IsDust() {
+			log.Infof("position: %+v", position)
+
+			side := types.SideTypeSell
+			takerPrice := fixedpoint.Zero
+
+			if position.IsShort() {
+				side = types.SideTypeBuy
+				takerPrice = ticker.Sell
+			} else if position.IsLong() {
+				side = types.SideTypeSell
+				takerPrice = ticker.Buy
+			}
+
+			orderForms = append(orderForms, types.SubmitOrder{
+				Symbol:      s.Symbol,
+				Type:        types.OrderTypeLimit,
+				Side:        side,
+				Price:       takerPrice,
+				Quantity:    position.GetQuantity(),
+				Market:      s.Market,
+				TimeInForce: types.TimeInForceGTC,
+				Tag:         "takeProfit",
+			})
+
+			log.Infof("SUBMIT TAKER ORDER: %+v", orderForms)
+
+			if _, err := s.Strategy.OrderExecutor.SubmitOrders(ctx, orderForms...); err != nil {
+				log.WithError(err).Error("unable to submit orders")
+			}
+
+			return
+		}
+
+		askQuantity = s.Market.AdjustQuantityByMinNotional(askQuantity, askPrice)
+		if !s.Market.IsDustQuantity(askQuantity, askPrice) && askQuantity.Compare(baseBalance.Available) < 0 {
```

Comment:
is the askQuantity need to mul `askPrice`?

Reply:
I believe the answer is no, only concerns whether quantity is sufficient in sell

Reply:
yeah, for base quantity we check it againt the base balance, so no conversion is needed